Parcelable issues in GameRequestContent

nodrock opened the pull request:
I modified GameRequestContent to properly handle parceling and also added setRecipients(List) instead of setTo(String) to make iOS and Android API as similar as it can be.

swiese commented:
Can you explain what you mean by "properly handle parceling"? Was there an issue?

nodrock commented:
Yes of course. 
1. When you parcel enums like this 
```java
out.writeString(this.getActionType().toString());
```
than it will throw null pointer exception when actionType is null. 
Also reading it like this is not safe and causes null pointer exception.
```java
this.actionType = ActionType.valueOf(in.readString());
```
There are two ways to handle this issue check it for null or make enums parcelable. I think the second option is better and more consistent.
2.
```java
this.suggestions = new ArrayList<>();
in.readStringList(this.suggestions);
```
Reading lists like this causes inconsistency if original suggestions variable was null it will be empty list after parceling. That causes problem in [GameRequestValidation](../blob/master/facebook/src/com/facebook/share/internal/GameRequestValidation.java) because suggestions are check against null and it is mutual exclusive with every other option. Proper way is using createStringArrayList.
```java
this.recipients = in.createStringArrayList();
```
3.
I changed 'to' property from string of comma-separated values to 'recipients' which is list of strings (similar to suggestions and consistent with FB iOS SDK).
4.
I don't like using string representation of enums like this 
```java
gameRequestContent.getActionType().toString().toLowerCase(Locale.ENGLISH));
```
It is not good for future changes. So I added value to ActionType and use getValue() instead of this.

Thats all I think. If you have any other question let me know. Thank you

ferlex commented:
Spy hacker's

On Sun, Sep 20, 2015, 4:07 AM nodrock <notifications@github.com> wrote:

> Yes of course.
> 1. When you parcel enums like this
>
> out.writeString(this.getActionType().toString());
>
> than it will throw null pointer exception when actionType is null.
> Also reading it like this is not safe and causes null pointer exception.
>
> this.actionType = ActionType.valueOf(in.readString());
>
> There are two ways to handle this issue check it for null or make enums
> parcelable. I think the second option is better and more consistent.
> 2.
>
> this.suggestions = new ArrayList<>();
> in.readStringList(this.suggestions);
>
> Reading lists like this causes inconsistency if original suggestions
> variable was null it will be empty list after parceling. That causes
> problem in GameRequestValidation
> <http://../blob/master/facebook/src/com/facebook/share/internal/GameRequestValidation.java>
> because suggestions are check against null and it is mutual exclusive with
> every other option. Proper way is using createStringArrayList.
>
> this.recipients = in.createStringArrayList();
>
> 3.
> I changed 'to' property from string of comma-separated values to
> 'recipients' which is list of strings (similar to suggestions and
> consistent with FB iOS SDK).
> 4.
> I don't like using string representation of enums like this
>
> gameRequestContent.getActionType().toString().toLowerCase(Locale.ENGLISH));
>
> It is not good for future changes. So I added value to ActionType and use
> getValue() instead of this.
>
> Thats all I think. If you have any other question let me know. Thank you
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/facebook-android-sdk/pull/445#issuecomment-141769552>
> .
>

swiese commented on facebook/src/com/facebook/internal/Utility.java:
```diff
@@ -498,7 +498,7 @@ public static void putNonEmptyString(Bundle b, String key, String value) {
     }
 
-    public static void putCommaSeparatedStringList(Bundle b, String key, ArrayList<String> list) {
+    public static void putCommaSeparatedStringList(Bundle b, String key, List<String> list) {
```
Nice!

swiese commented on facebook/src/com/facebook/share/model/GameRequestContent.java:
```diff
@@ -21,37 +21,101 @@
-        SEND,
-        ASKFOR,
-        TURN,
+    public enum ActionType implements Parcelable {
```
Instead of making this parcelable, which adds a lot of boilerplate code, what about doing what we do in [AppGroupCreationContent](
https://github.com/facebook/facebook-android-sdk/blob/001984c85bbb5ccb7a3df4e33fe3090d22a8ac88/facebook/src/com/facebook/share/model/AppGroupCreationContent.java#L42)

And just add them as serializable?

nodrock replied:
I agree it will be more readable. I will modify it.

swiese commented on facebook/src/com/facebook/share/model/GameRequestContent.java:
```diff
@@ -21,37 +21,101 @@
 
     private final String message;
-    private final String to;
+    private final List<String> recipients;
```
Nice call on adding recipients. This is a much nicer api that is far less error prone

swiese commented:
Thanks for another contribution!

funset commented:
i hope to know what param for GameRequestContent.Builder().setRecipients() ,can you help me ,thinks~

mingflifb commented:
It's a list of app-scoped user ids.

funset commented:
ok, but how can i get selcted  user id and user name from defualt  friend selector when my sent a  game request
